java/src/main/java/ai/onnxruntime/OrtSession.java

```diff
@@ -401,4 +389,101 @@ public void addNuphar(boolean allowUnalignedBuffers, String settings) throws Ort
         private native void addNnapi(long apiHandle, long nativeHandle) throws OrtException;
         private native void addNuphar(long apiHandle, long nativeHandle, int allowUnalignedBuffers, String settings) throws OrtException;
     }
+
+    /**
+     * An {@link AutoCloseable} wrapper around a {@link Map} containing {@link OnnxValue}s.
+     * <p>
+     * When this is closed it closes all the {@link OnnxValue}s inside it. If you maintain a reference to a
+     * value after this object has been closed it will throw an exception upon access.
```

Comment:
That sounds kind of limiting and is exactly the kind of thing we can do with reference counting to achieve acceptable levels of performance, but at least this issue is well documented. :+1:  

@EronWright Do you feel comfortable with TensorFlow also working like this? If not, speak now or this is how Adam will make it work there as well :)

Reply:
I'm not planning to do this to the TF JVM API.

Reply:
Why not? I thought you would have welcomed the consistency...

Reply:
Because this was the consensus on how to deal with this issue here, and in the TF JVM group there are different people involved who have different opinions. Plus the scoping issues for TF are larger as it allows model training, so tensors have a variety of lifetimes.

Reply:
Ok, I assume you had a conversation about that offline. According to the last comment at https://github.com/microsoft/onnxruntime/pull/2215#discussion_r341372996 it sounded to me like the idea was to do something very similar to `PointerScope` in JavaCPP. That is, you have your collection of something, and then you have a scope object, which contains those objects, maybe as a reference to the collection, but it doesn't give you access to those objects. It's just there to close them, and maybe also allow adding some more or remove some of them.

Reply:
It's pretty similar to a scope, but as it's only returned by a single method, and adding things to this scope would be pointless, it doesn't have any additional methods. The only thing missing is a way to remove an `OnnxValue` from the `Result` scope, and I'm unsure of the benefit of that. We can discuss that more after people have used the API, if it becomes necessary then it's not too hard to change it.